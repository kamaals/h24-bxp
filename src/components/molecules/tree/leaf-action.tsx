"use client";
import { Button } from "@/components/atoms/button";
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuGroup,
  DropdownMenuItem,
  DropdownMenuLabel,
  DropdownMenuSeparator,
  DropdownMenuTrigger,
} from "@/components/atoms/dropdown-menu";
import { EllipsisVertical, Plus, Edit, List } from "lucide-react";
import ActionConfirm from "@/components/molecules/action-confirm/action-confirm";
import React from "react";

type Props = {
  addAction?: () => void;
  deleteAction?: () => void;
  editAction?: () => void;
  addProduct?: () => void;
  showProducts?: () => void;
};

export function LeafAction({
  addAction,
  deleteAction,
  editAction,
  addProduct,
  showProducts,
}: Props) {
  const [isOpen, setOpen] = React.useState(false);

  return (
    <DropdownMenu open={isOpen} onOpenChange={setOpen}>
      <DropdownMenuTrigger asChild>
        <Button
          data-testid="leaf-action-toggle"
          type="button"
          size="sm"
          variant="ghost"
        >
          <EllipsisVertical />
        </Button>
      </DropdownMenuTrigger>
      <DropdownMenuContent
        data-testid="leaf-action-dropdown"
        align={"end"}
        className="w-56"
      >
        <DropdownMenuLabel>Actions</DropdownMenuLabel>
        <DropdownMenuSeparator />
        <DropdownMenuGroup>
          <DropdownMenuItem
            data-testid={"leaf-action-dropdown-add"}
            onClick={addAction}
          >
            <Plus /> Add
          </DropdownMenuItem>
          <DropdownMenuItem
            data-testid={"leaf-action-dropdown-edit"}
            onClick={editAction}
          >
            <Edit /> Edit
          </DropdownMenuItem>
          <div className="relative flex cursor-default select-none items-center gap-2 rounded-sm text-sm outline-none transition-colors hover:bg-accent hover:text-accent-foreground focus:bg-accent focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50 [&>svg]:size-4 [&>svg]:shrink-0">
            <ActionConfirm onConfirm={deleteAction} />
          </div>
        </DropdownMenuGroup>
        <DropdownMenuSeparator />
        <DropdownMenuGroup>
          <DropdownMenuItem
            onClick={addProduct}
            data-testid={"leaf-action-dropdown-add-product"}
          >
            <Plus /> Add Product
          </DropdownMenuItem>
        </DropdownMenuGroup>
        <DropdownMenuGroup>
          <DropdownMenuItem
            onClick={showProducts}
            data-testid={"leaf-action-dropdown-add-product"}
          >
            <List /> View Products
          </DropdownMenuItem>
        </DropdownMenuGroup>
      </DropdownMenuContent>
    </DropdownMenu>
  );
}
