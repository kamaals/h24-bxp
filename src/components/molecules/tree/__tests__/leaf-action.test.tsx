import { describe, it, expect, beforeEach } from "vitest";
import { render, screen, cleanup, waitFor } from "@testing-library/react";
import { LeafAction } from "../leaf-action";
import userEvent from "@testing-library/user-event";

describe(LeafAction.name, async () => {
  beforeEach(() => {
    cleanup();
  });
  it("Should match the snapshot", () => {
    const resp = render(<LeafAction />);
    expect(resp).toMatchSnapshot();
  });

  it("Should match the snapshot", async () => {
    const user = userEvent.setup();
    render(<LeafAction />);
    const actionToggle = screen.getByTestId("leaf-action-toggle");
    await user.click(actionToggle);
    await waitFor(() => {
      expect(screen.getByTestId("leaf-action-dropdown")).toBeInTheDocument();
    });
  });
});
