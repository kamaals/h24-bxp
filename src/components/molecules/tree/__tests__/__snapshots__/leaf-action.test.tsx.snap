// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LeafAction > Should match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5"
        data-slot="dropdown-menu-trigger"
        data-state="closed"
        data-testid="leaf-action-toggle"
        id="radix-«r0»"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-ellipsis-vertical"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="1"
          />
          <circle
            cx="12"
            cy="5"
            r="1"
          />
          <circle
            cx="12"
            cy="19"
            r="1"
          />
        </svg>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-expanded="false"
      aria-haspopup="menu"
      class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5"
      data-slot="dropdown-menu-trigger"
      data-state="closed"
      data-testid="leaf-action-toggle"
      id="radix-«r0»"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-ellipsis-vertical"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="1"
        />
        <circle
          cx="12"
          cy="5"
          r="1"
        />
        <circle
          cx="12"
          cy="19"
          r="1"
        />
      </svg>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
